Fix Issue 16140 -  while(auto x = y) does not behave like if(auto x = y)

RazvanN7 opened the pull request:
This includes:

 - make parser recognize `while(auto a = exp)`
 - add a field to `WhileStatement` to keep track of the defined variable
 - ~~rewrite the `while` loop to `for(auto a = exp; a; a = exp)`~~
 - update C++ headers
 - rewrite `while(auto a = exp)` to `while(true) if (auto a = exp) {/* loop body */ } else break;` as suggested by @schveiguy (thanks!). ~~The lowering is handled in the parser.~~ The lowering takes place during semantic analysis.
 - Update `hdrgen.d` to properly output the introduced `while` version.
 - Update `astbase.d`


I think a changelog entry might be necessary. What do you think?

RazvanN7 committed:
Fix Issue 16140 - while(auto x = y) does not behave like if(auto x = y)

RazvanN7 commented on src/dmd/parse.d:
```diff
@@ -5528,6 +5528,90 @@ final class Parser(AST) : Lexer
+    /***
+     * Parse an assignment condition for if or while statements.
+     *
+     * Params:
```
Leftover from a previous version. Need to update to `Returns`, but waiting to see what the testsuite thinks.

RazvanN7 commented on src/dmd/statement.d:
```diff
@@ -1155,13 +1155,15 @@ extern (C++) final class ForwardingStatement : Statement
     Loc endloc;             // location of closing curly bracket
 
-    extern (D) this(const ref Loc loc, Expression condition, Statement _body, Loc endloc)
+    extern (D) this(const ref Loc loc, Parameter prm, Expression condition, Statement _body, Loc endloc)
```
This needs to be a different overload as to not break ldc/gdc.

deadalnix commented on src/dmd/parse.d:
```diff
@@ -5951,79 +6037,7 @@ final class Parser(AST) : Lexer
-                }
-                else if (storageClass != 0)
-                    error("found `%s` while expecting `=` or identifier", n.toChars());
-
```
Awesome negative coding is action!

RazvanN7 replied:
That's actually more of a refactoring :)). The same code will be used for `with` also.

TungstenHeart commented:
You need to test `while (StorageClass Type = expression)` too

andre2007 commented:
@RazvanN7 fantastic that you work on this topic. I assume after this, the only missing place is `with(auto ...)`.

MoonlightSentinel commented:
> * rewrite the `while` loop to `for(auto a = exp; a; a = exp)`

As pointed out by @schveiguy on slack, this lowering causes a missmatch between the first and all remaining iterations (construction vs assignment).

Maybe this should rather be rewritten as

```d
while (true)
{
    auto a = exp;
    if (!a) break;

   ....
}
```

deadalnix commented:
Looking at what clang does, it implements a crazy state machine to handle this. See https://godbolt.org/z/c6E138 for instance.

But indeed, the declaration happens INSIDE the loop. See https://godbolt.org/z/3WM3YT vs https://godbolt.org/z/9vzKPT .

This lowering is wrong.

ntrel commented:
> I assume after this, the only missing place is with(auto ...)

There's also `switch (auto v = cond)` ;-)

>> for(auto a = exp; a; a = exp)

> this lowering causes a mismatch between the first and all remaining iterations

Yes, it wouldn't work for certain types that can't be assigned. The `while(true)` lowering would work.

schveiguy commented:
Another possible rewrite:

```D
while(true)
   if(auto a = exp) { 
      /* loop body */
   else break;
```

p0nce commented:
Reading such code, I can't help but to wonder what the lifetime of the inner variable is. Does it get destroyed after each iteration? One for all iteration? Does seem readable and self-evident to me.

BorisCarvajal commented:
I think for consistency and ease of reasoning, the behavior should be the same as C++.

The summary and potential lowering are described here:
https://en.cppreference.com/w/cpp/language/while

schveiguy commented:
> I think for consistency and ease of reasoning, the behavior should be the same as C++.

Yeah, that looks good actually.

WalterBright commented:
> Yeah, that looks good actually.

It looks the same as your suggestion.

WalterBright reviewed (dismissed):
hdrgen.d needs to properly print it out.

RazvanN7 commented:
@WalterBright @deadalnix @MoonlightSentinel I have updated the code with the requested changes. Please give this another pass.

I have added a brittle test for `hdrgen.d` to recognize the lowered `WhileStatement` : if we have a `while(true) if (auto v  = exp)` and the end location of the underlying `IfStatement` is the default one, then it is a lowered `WhileStatement`. This should suffice, as user defined `while` statements should always have location. I think that this hack is worth it, otherwise a field should be added to `WhileStatement` just to track the variable for header generation.

ibuclaw commented on src/dmd/hdrgen.d:
```diff
@@ -250,11 +250,47 @@ public:
+         * while(true)
+         *     if (auto a = exp) { ... }
+         *
+         * It is possible that this code was lowered
```
Foreach statements don't have this problem, do they?

RazvanN7 replied:
Foreach statements have a `Parameter` array that stores all the initialization variables, therefore `hdrgen` can print them directly. Those parameters are needed inside when the `foreach` is lowered to a `for` loop. I can do the same, but that would require storing the parameter in the `WhileStatement` - with the current implementation that would wasteful since generating headers occurs less often than normal compilation.

ibuclaw commented on src/dmd/parse.d:
```diff
@@ -5851,13 +5935,36 @@ final class Parser(AST) : Lexer
+                    // it is set to an initial value so that the
+                    // header generator recognizes this lowering
+                    AST.IfStatement ifd = new AST.IfStatement(loc, param, condition, _body, new AST.BreakStatement(loc, null), Loc());
+                    s = new AST.WhileStatement(loc, new AST.IntegerExp(loc, 1, AST.Type.tbool), ifd, endloc);
```
Can't this rewrite wait until the semantic pass? For instance, what if there are semantic errors in the condition (such as unresolved symbols)?

RazvanN7 replied:
I don't see any benefit in waiting up until then. We would need to store `param` for that in the `WhileStatement` and what would that bring us?

To answer your question: the `if` condition is initialized with the start location of the `while` statement, therefore any errors that would occur would be treated the same, except that they are in child scope of the `while` loop. This should not affect in any way error reporting.

UplinkCoder replied:
Please don't lower in the parser.

RazvanN7 replied:
I can do the lowering in the semantic phase, but that will require the addition of a field. That will make things easier for `hdrgen`, but that will make all instances of `WhileStatement` larger.

ibuclaw replied:
@RazvanN7 - yeah, I was just mindful of the possibility of there being IfStatement-specific error messages that would only serve to confuse the user.

Another example are clients that use the DMD front-end to get symbolic information out of source files (e.g: Editors?).

ibuclaw replied:
> I can do the lowering in the semantic phase, but that will require the addition of a field. That will make things easier for `hdrgen`, but that will make all instances of `WhileStatement` larger.

I don't think size is a good argument against.  `WhileStatement`'s aren't that important in the bid to keep memory footprint down, unlike `Loc` or `TemplateInstance`.

BorisCarvajal replied:
Maybe add the required field to `Statement` it seems it has some space available. And it could serve other purposes too.

UplinkCoder commented on src/dmd/hdrgen.d:
```diff
@@ -250,11 +250,47 @@ public:
+        {
+            if (cond.isBool(true))
+            {
+                auto ifd = s._body.isIfStatement();
```
why is it called `ifd` 
what does `ifd` stand for?

UplinkCoder reviewed (dismissed):
Please don't do lowering in the parser.
it's so confusing.

Geod24 commented:
@RazvanN7 : I also agree we shouldn't lower in the parser. At this point, I would look into unifying `ForStatement` and `WhileStatement` under the same banner, since they will end up having almost exactly the same semantic.

RazvanN7 commented:
@Geod24 @UplinkCoder @ibuclaw @WalterBright I gave this another shot. I added a `Parameter` to `WhileStatement` and performed the lowering in the semantic analysis phase.

RazvanN7 commented:
I have added a changelog entry and made a spec PR: https://github.com/dlang/dlang.org/pull/2977

RazvanN7 commented on test/runnable/test16140.d:
```diff
@@ -0,0 +1,32 @@
+    while(const value = fun())
+        res ~= value;
+    assert(res == [1, 2, 3]);
+}
```
I don't think this is necessary as this patch builds on the `if(auto n = exp)` condition logic that is heavily tested right now.

Geod24 commented on changelog/while-condition-assignment.dd:
```diff
@@ -0,0 +1,17 @@
+same semantics as:
+
+---
+while(true)
```
```suggestion
while (true)
```

Geod24 commented on src/dmd/statementsem.d:
```diff
@@ -504,7 +504,24 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
+             *     { break; }
+             */
+            _body = new IfStatement(ws.loc, ws.param, ws.condition, ws._body, new BreakStatement(ws.loc, null), ws.endloc);
+            cond = new IntegerExp(ws.loc, 1, Type.tbool);
```
Don't we have a statically allocated constant for this ?

Geod24 commented on test/compilable/extra-files/header2.d:
```diff
@@ -162,6 +162,17 @@ align(2) struct S12200_2
+void gun()()
+{
+    int[] res;
+    while(auto va = fun()) {}  // expression expected, not 'auto'
```
```suggestion
    while (auto va = fun()) {}
```

Because 5 years from now, the former error message would just confuse readers

Geod24 commented on test/compilable/extra-files/header2.d:
```diff
@@ -162,6 +162,17 @@ align(2) struct S12200_2
+    int[] res;
+    while(auto va = fun()) {}  // expression expected, not 'auto'
+
+    while(true)
```
```suggestion
    while (true)
```

Geod24 commented on src/dmd/hdrgen.d:
```diff
@@ -250,6 +250,20 @@ public:
+                buf.writeByte(' ');
+            if (p.type)
+                typeToBuffer(p.type, p.ident, buf, hgs);
+            else
+                buf.writestring(p.ident.toString());
```
This else looks wrong. It should print the name unconditionally, and add an extra parenthesis after `typeToBuffer`. Try `while (scope int x = foo())` to see the bug.

RazvanN7 replied:
It's not wrong. You can have: `while (int a = fun())` in which case the `if` is taken and it prints the type and the identifier (as it is passed as a parameter and typeToBuffer prints it if is non-null). Or you can have `while (const a = fun())` in which case the else branch is taken. This the correct behavior and it is exactly the same as the `if` one.

Geod24 reviewed (changes requested):
Nice and simple, just a bug on header printing

RazvanN7 commented:
@Geod24 I have addressed your comments.
